src/inmanta/protocol/rest/__init__.py

```diff
@@ -219,23 +219,39 @@ async def process(self) -> None:
         for i, arg in enumerate(args):
             arg_type: Optional[type[object]] = self._argspec.annotations.get(arg)
             if arg in self._message:
-                # Argument is parameter in body of path of HTTP request
-                if (
-                    arg_type
-                    and self._properties.operation == "GET"
-                    and typing_inspect.is_generic_type(arg_type)
-                    and issubclass(typing_inspect.get_origin(arg_type), list)
-                    and not isinstance(self._message[arg], list)
-                    and len(typing_inspect.get_args(arg_type, evaluate=True)) == 1
-                    and isinstance(self._message[arg], typing_inspect.get_args(arg_type)[0])
-                ):
-                    # If a GET endpoint has a parameter of type list that is encoded as a URL query parameter and the
-                    # specific request provides a list with one element, urllib doesn't parse it as a list.
-                    # Map it here explicitly to a list.
-                    value = [self._message[arg]]
+                message_arg = self._message[arg]
+
+                if self._properties.operation == "GET":
+                    # Check if the argument type is an Optional type (e.g., Optional[List[SomeType]])
+                    if typing_inspect.is_optional_type(arg_type):
+                        # Extract non-None types from the Optional type
+                        non_none_arg_types = [arg for arg in typing_inspect.get_args(arg_type) if arg is not type(None)]
+
+                        # If there's only one type and it's a list, simplify the type to just the list type
+                        if len(non_none_arg_types) == 1 and typing_inspect.get_origin(non_none_arg_types[0]) is list:
+                            arg_type = non_none_arg_types[0]
+
+                    # Check if the argument type is a generic list
+                    is_generic_list = (
+                        arg_type and typing_inspect.is_generic_type(arg_type) and typing_inspect.get_origin(arg_type) is list
+                    )
+                    # Check if the list type has exactly one element type
+                    is_single_element_list = len(typing_inspect.get_args(arg_type, evaluate=True)) == 1
+                    # Check if the argument value itself is not a list
+                    is_not_list = not isinstance(self._message[arg], list)
+
+                    if is_generic_list and is_not_list and is_single_element_list:
+                        # If a GET endpoint has a parameter of type list that is encoded as a URL query parameter and the
+                        # specific request provides a list with one element, urllib doesn't parse it as a list.
+                        # Map it here explicitly to a list.
+                        value = [self._message[arg]]
```

Comment:
The problem with this new implementation is that the goal is rather vague. In the past we had this comment that clarified a lot. But with the new implementation this comment has sunk quite deep into the if-the-else nesting. I think readability would be improved quite a lot if we extract this logic into a separate method. Then we can put this comment in the docstring for example.